.snippets/code/store-and-retrieve-data/use-storagehub-sdk/end-to-end-storage-workflow/fileOperations.ts

```diff
@@ -238,3 +239,111 @@ export async function verifyDownload(
+        i + 1
+      } of ${maxAttempts}...`
+    );
+
+    // Query the runtime for the StorageRequest entry associated with this fileKey
+    const req = await polkadotApi.query.fileSystem.storageRequests(fileKey);
+
+    // StorageRequest removed from state before confirmation is an error
+    if (req.isNone) {
+      throw new Error(
+        `StorageRequest for ${fileKey} no longer exists on-chain.`
+      );
+    }
+
+    // Decode the on-chain metadata struct
+    const data: PalletFileSystemStorageRequestMetadata = req.unwrap();
+
+    // Extract the MSP confirmation tuple (mspId, bool)
+    const mspTuple = data.msp.isSome ? data.msp.unwrap() : null;
+
+    // The second value in the tuple is a SCALE Bool (codec), so convert using .isTrue
+    const mspConfirmed = mspTuple ? (mspTuple[1] as any).isTrue : false;
+
+    // If MSP has confirmed the storage request, we’re good to proceed
+    if (mspConfirmed) {
+      console.log('Storage request confirmed by MSP on-chain');
+      return;
+    }
+
+    // Wait before polling again
+    await new Promise((r) => setTimeout(r, delayMs));
+  }
+
+  // All attempts exhausted
+  throw new Error(
+    `FileKey ${fileKey} not ready for download after waiting ${
+      maxAttempts * delayMs
+    } ms`
+  );
+}
+// --8<-- [end:wait-for-msp-confirm-on-chain]
+
+// --8<-- [start:wait-for-backend-file-ready]
+export async function waitForBackendFileReady(
+  bucketId: string,
+  fileKey: string
+) {
+  const maxAttempts = 15; // Number of polling attempts
+  const delayMs = 2000; // Delay between attempts in milliseconds
+
+  for (let i = 0; i < maxAttempts; i++) {
+    console.log(
+      `Checking for file in MSP backend, attempt ${i + 1} of ${maxAttempts}...`
+    );
+
+    try {
+      // Query MSP backend for the file metadata
+      const fileInfo = await mspClient.files.getFileInfo(bucketId, fileKey);
+
+      // File is fully ready — backend has indexed it and can serve it
+      if (fileInfo.status === 'ready') {
+        console.log('File found in MSP backend:', fileInfo);
+        return fileInfo;
+      }
+
+      // Failure statuses (irrecoverable for this upload lifecycle)
+      if (fileInfo.status === 'revoked') {
+        throw new Error('File upload was cancelled by user');
+      } else if (fileInfo.status === 'rejected') {
+        throw new Error('File upload was rejected by MSP');
+      } else if (fileInfo.status === 'expired') {
+        throw new Error('File upload request expired before MSP processed it');
+      }
+
+      // Otherwise still pending (indexer not done, MSP still syncing, etc.)
+      console.log(`File status is "${fileInfo.status}", waiting...`);
+    } catch (error: any) {
+      if (error?.status === 404 || error?.body?.error === 'Not found: Record') {
+        // Handle "not yet indexed" as a *non-fatal* condition
+        console.log(
+          'File not yet indexed in MSP backend (404 Not Found). Waiting before retry...'
```

Comment:
My AI Sidekick found this potential problem: 
`The catch in waitForBackendFileReady always executes the unconditional log/throw after the 404/expected-error branch, so the very first “not indexed yet” 404 will throw and abort instead of retrying. The loop never actually polls; uploads followed by download will fail on first pass. Needs an early continue/return to avoid throwing on handled 404s.`